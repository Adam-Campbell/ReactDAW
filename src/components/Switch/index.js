import React, { Component } from 'react';
import PropTypes from 'prop-types';

const Switch = props => (
    <React.Fragment>
        <p className="switch__row-description">{props.rowDescription}</p>
            <div className="switch__row">
                {props.optionsData.map((option, index) => (
                    <React.Fragment
                        key={option.id}
                    >
                        <input 
                            className="switch__radio-button" 
                            type="radio" 
                            id={option.id} 
                            name={option.name}
                            value={option.value} 
                            onChange={(e) => props.handleChange(e.target.value)} 
                            checked={props.value === option.value}
                        ></input>
                        <label 
                            className="switch__label" 
                            htmlFor={option.id}
                        >{option.text}</label>
                    </React.Fragment>
                ))}
            </div>
    </React.Fragment>
);

Switch.propTypes = {
    value: PropTypes.string.isRequired,
    handleChange: PropTypes.func.isRequired,
    rowDescription: PropTypes.string.isRequired,
    optionsData: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        value: PropTypes.string,
        id: PropTypes.string.isRequired,
        text: PropTypes.string.isRequired
    })).isRequired 
}

export default Switch;
