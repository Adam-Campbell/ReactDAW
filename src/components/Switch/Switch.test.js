import React from 'react';
import { shallow } from 'enzyme';
import Switch from './';

test('renders correctly', () => {
    const wrapper = shallow(
        <Switch 
            value="option-one"
            handleChange={jest.fn()}
            rowDescription="Switch"
            optionsData={[
                { name: 'options', value: 'option-one', id: 'option-one', text: 'Option One' },
                { name: 'options', value: 'option-two', id: 'option-two', text: 'Option Two' },
            ]}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('a radio button and label are created for each option object passed in as a prop', () => {
    const wrapper = shallow(
        <Switch 
            value="option-one"
            handleChange={jest.fn()}
            rowDescription="Switch"
            optionsData={[
                { name: 'options', value: 'option-one', id: 'option-one', text: 'Option One' },
                { name: 'options', value: 'option-two', id: 'option-two', text: 'Option Two' },
            ]}
        />
    );
    const radioButtonsWrapper = wrapper.find('.switch__radio-button');
    const labelsWrapper = wrapper.find('.switch__label');
    expect(radioButtonsWrapper).toHaveLength(2);
    expect(labelsWrapper).toHaveLength(2);
    expect(radioButtonsWrapper.at(0).prop('id')).toBe('option-one');
    expect(radioButtonsWrapper.at(1).prop('id')).toBe('option-two');
    expect(labelsWrapper.at(0).prop('htmlFor')).toBe('option-one');
    expect(labelsWrapper.at(1).prop('htmlFor')).toBe('option-two');
});

test('handleChange is called when a change event is fired', () => {
    const mockedHandleChange = jest.fn();
    const wrapper = shallow(
        <Switch 
            value="option-one"
            handleChange={mockedHandleChange}
            rowDescription="Switch"
            optionsData={[
                { name: 'options', value: 'option-one', id: 'option-one', text: 'Option One' },
                { name: 'options', value: 'option-two', id: 'option-two', text: 'Option Two' },
            ]}
        />
    );
    expect(mockedHandleChange).toHaveBeenCalledTimes(0);
    wrapper.find('.switch__radio-button').at(1).simulate('change', { target: { value: 'option-two' } });
    expect(mockedHandleChange).toHaveBeenCalledTimes(1);
    expect(mockedHandleChange.mock.calls[0][0]).toBe('option-two');
});