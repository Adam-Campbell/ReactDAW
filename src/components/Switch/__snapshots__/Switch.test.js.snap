// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <p
    className="switch__row-description"
  >
    Switch
  </p>
  <div
    className="switch__row"
  >
    <input
      checked={true}
      className="switch__radio-button"
      id="option-one"
      name="options"
      onChange={[Function]}
      type="radio"
      value="option-one"
    />
    <label
      className="switch__label"
      htmlFor="option-one"
    >
      Option One
    </label>
    <input
      checked={false}
      className="switch__radio-button"
      id="option-two"
      name="options"
      onChange={[Function]}
      type="radio"
      value="option-two"
    />
    <label
      className="switch__label"
      htmlFor="option-two"
    >
      Option Two
    </label>
  </div>
</Fragment>
`;
